File: Examples/elixir/basics.exs
defmodule Math do
  def fib(0), do: 0
  def fib(1), do: 1
  def fib(n) when n > 1, do: fib(n - 1) + fib(n - 2)

  def square(x), do: x * x

  def sum(list) do
    Enum.reduce(list, 0, fn x, acc -> x + acc end)
  end

  def classify(n) do
    cond do
      n < 0 -> "negative"
      n == 0 -> "zero"
      n < 10 -> "small"
      true -> "large"
    end
  end
end

fibs = Enum.map(0..9, &Math.fib/1)
IO.puts("fib: #{Enum.join(fibs, ", ")}")

nums = [5, 3, 9, 1, 7]
IO.puts("sum: #{Math.sum(nums)}")
IO.puts("squares: #{Enum.join(Enum.map(nums, &Math.square/1), ", ")}")
IO.puts("big: #{Enum.join(Enum.filter(nums, fn x -> x > 3 end), ", ")}")
IO.puts("sorted: #{Enum.join(Enum.sort(nums), ", ")}")
IO.puts("count: #{Enum.count(nums)}")

piped = nums |> Enum.filter(fn x -> x > 3 end) |> Enum.map(fn x -> x * 10 end) |> Enum.sum()
IO.puts("piped: #{piped}")

for n <- [-5, 0, 3, 42] do
  IO.puts("#{n} is #{Math.classify(n)}")
end

point = {3, 4}
{x, y} = point
IO.puts("point: #{x}, #{y}")

person = %{name: "Alice", age: 30}
IO.puts("name: #{person.name}, age: #{Map.get(person, :age)}")
IO.puts("keys: #{Enum.join(Enum.map(Map.keys(person), &to_string/1), ", ")}")

text = "Hello, Elixir"
IO.puts("upper: #{String.upcase(text)}")
IO.puts("length: #{String.length(text)}")
IO.puts("parts: #{Enum.join(String.split(text, ", "), "|")}")

result = case 7 do
  1 -> "one"
  7 -> "seven"
  _ -> "other"
end
IO.puts("case: #{result}")

value = if 3 > 2, do: "yes", else: "no"
IO.puts("if: #{value}")

IO.puts("concat: #{"foo" <> "bar"}")
IO.puts("list: #{Enum.join([1, 2] ++ [3, 4], ",")}")
IO.puts("div: #{div(17, 5)} rem: #{rem(17, 5)}")
IO.inspect(%{a: 1, b: 2})
IO.inspect([1, :two, "three"])

File: Examples/erlang/basics.erl
-module(demo).
-export([main/0]).

fib(0) -> 0;
fib(1) -> 1;
fib(N) when N > 1 -> fib(N - 1) + fib(N - 2).

square(X) -> X * X.

classify(N) when N < 0 -> negative;
classify(0) -> zero;
classify(N) when N < 10 -> small;
classify(_) -> large.

describe({point, X, Y}) -> io:format("point ~p,~p~n", [X, Y]);
describe({circle, R}) -> io:format("circle r=~p~n", [R]);
describe(_) -> io:format("unknown~n", []).

sum([]) -> 0;
sum([H|T]) -> H + sum(T).

main() ->
    Fibs = lists:map(fun fib/1, lists:seq(0, 9)),
    io:format("fib: ~p~n", [Fibs]),

    Nums = [5, 3, 9, 1, 7],
    io:format("sum: ~p~n", [sum(Nums)]),
    io:format("lists:sum: ~p~n", [lists:sum(Nums)]),
    io:format("squares: ~p~n", [lists:map(fun square/1, Nums)]),
    io:format("big: ~p~n", [lists:filter(fun(X) -> X > 3 end, Nums)]),
    io:format("sorted: ~p~n", [lists:sort(Nums)]),
    io:format("reversed: ~p~n", [lists:reverse(Nums)]),
    io:format("length: ~p~n", [length(Nums)]),
    io:format("folded: ~p~n", [lists:foldl(fun(X, Acc) -> X + Acc end, 0, Nums)]),

    lists:map(fun(N) -> io:format("~p is ~p~n", [N, classify(N)]) end, [-5, 0, 3, 42]),

    describe({point, 3, 4}),
    describe({circle, 2}),
    describe(other),

    Point = {3, 4},
    {X, Y} = Point,
    io:format("x=~p y=~p~n", [X, Y]),

    Person = #{name => "Alice", age => 30},
    io:format("name: ~s age: ~p~n", [maps:get(name, Person), maps:get(age, Person)]),

    Result = case 7 of
        1 -> one;
        7 -> seven;
        _ -> other
    end,
    io:format("case: ~p~n", [Result]),

    Value = if
        3 > 2 -> yes;
        true -> no
    end,
    io:format("if: ~p~n", [Value]),

    io:format("concat: ~p~n", [[1, 2] ++ [3, 4]]),
    io:format("div: ~p rem: ~p~n", [17 div 5, 17 rem 5]),
    io:format("upper: ~s~n", [string:to_upper("hello erlang")]),
    io:format("joined: ~s~n", [string:join(["a", "b", "c"], "-")]).
